### PR TITLE
feat: Improved localization API

### LINKED ISSUE #327: Improved localization APIs
### Feature Request

Instead of using `browser.i18n.getMessage` directly, include a custom `i18n` API that can be used in it's place.

It should:

- Read localizations from `<srcDir>/locales/[country-code].(json|json5|yaml|yml)`
- Generate correctly formatted `.output/[target]/_locales/[country-code]/messages.json` files
- Support plural forms
- Support interpolation

Basically, for message files that look like this:

```yml
# locales/en.yml
simple: Hello world!
manifestStyle:
  message: This is the translated text
  description: This is a description to give translators more context (it's never translated)
namedInterpolation: Hello $name1$, my name is $name2$
arrayInterpolation: Hello $1, my name is $2
some:
  nested:
    translation: You can nest translations inside deep objects
escapeTheDollarSign: You owe me $$100
pluralForm:
  0: Zero items
  1: 1 item
  n: $count$ items
```

Would be translated to look like this:

```json5
// .output/_locales/en.json
{
  "simple": {
    "message": "Hello world!"
  },
  "manifestStyle": {
    "message": "This is the translated text",
    "description": "This is a description to give translators more context (it's never translated)"
  },
  "namedInterpolation": {
    "message": "Hello $name1$, my name is $name2$",
    "placeholders": {
      "name1": {
        "content": "$1"
      },
      "name2": {
        "content": "$2"
      }
    }
  },
  "arrayInterpolation": {
    "message": "Hello $1, my name is $2"
  },
  "some_nested_translation": {
    "message": "You can nest translations inside deep objects"
  },
  "escapeTheDollarSign": {
    "message": "You owe me $$100"
  },
  "pluralForm": {
    "message": "Zero items | 1 item | $count$ items",
    "placeholders": {
      "count": {
        "content": "$1"
      }
    }
  }
}
```

The API would look something like this:

```ts
// Auto-imported from:
// import { createI18n } from 'wxt/i18n';

const i18n = createI18n();

i18n.t("simple");                                             // Hello world!
i18n.t("manifestStyle");                                      // This is the translated text
i18n.t("namedInterpolation", { name1: "John", name2: "Joe" }) // Hello John, my name is Joe
i18n.t("arrayInterpolation", ["John", "Joe"])                 // Hello John, my name is Joe
i18n.t("some.nested.translation");                            // You can nest translations inside deep objects
i18n.t("escapeTheDollarSign");                                // You owe me $100
i18n.t("pluralForm", { count: 0 });                           // Zero items
i18n.t("pluralForm", { count: 1 });                           // 1 item
i18n.t("pluralForm", { count: 4 });                           // 4 items
```

### Acceptance Criteria

- [ ] Currently, we generate types for `browser.i18n.getMessage` to make sure the keys passed in are type-safe. Let's keep this logic, but how it is generated will be different.
   https://github.com/wxt-dev/wxt/blob/ab6a34afb690b0c39b9ffa3896f3cf333441a421/src/core/utils/building/generate-wxt-dir.ts#L96-L98
- [ ] Not everyone will be happy with the name "locales" for the new directory. We'll use that as a default, but let's add `localesDir` to the config to allow customization
- [ ] After copying public assets, we'll need to read in all files in `<srcDir>/<localesDir>/*`, transform them, and output them to `.output/[target]/_locales/[country-code]/messages.json`. Public assets are copied here:
   https://github.com/wxt-dev/wxt/blob/ab6a34afb690b0c39b9ffa3896f3cf333441a421/src/core/utils/building/build-entrypoints.ts#L28
- [ ] `i18n.t`
   - [ ] Plural form support
      - [ ] Use a special named placeholder, `count`, to detect keys that are plural
   - [ ] Interpolation
   - [ ] Escape `$` correctly

### What are the alternatives?

See #217 for a discussion of alternative methods.

### Additional context

Created based on https://github.com/wxt-dev/wxt/issues/217#issuecomment-1879788065

### PR DESCRIPTION
**Pull request by aklinker1**
This closes #327 . It adds a new module (that's auto-imported by default), `wxt/i18n`. To use the API:

```yml
# locales/en.yml
hello: Hello $1
items:
  1: 1 item
  n: $1 items
```

```ts
import { i18n } from 'wxt/i18n';

i18n.t("hello", ["Mark"]); // "Hello Mark"
i18n.tp("items", 2, ["1"]) // "2 items"
i18n.tp("items", 1, ["1"]) // "1 item"
```

- [x] Implement conversion tools for mapping message files
- [x] Implement client code
   - [x] Plural support
      - [x] Double check if plural logic works as expected for definitions without 0 or 1, just n
   - [x] Substitutions
   - [x] Overridable types
   - [x] Non-overridden types work
   - ~~_Typed placeholders_~~ Will add this in a future PR, but the overidable types support defining types for the placeholders.
- [x] Update type generation for `browser.i18n.getMessage`
- [x] Generate types for new `i18n` variable
- [x] Update tests
- [x] Update docs
- [x] Release test version
- [ ] Create PR to update i18n examples
- [ ] Check compatibility with i18next VS code plugin

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "wxt",
   "type": "module",
-  "version": "0.14.1",
+  "version": "0.14.2-alpha1",
```
Comment (aklinker1) [inline package.json:4]: Revert before merging

**Comment by aklinker1** [comment]
@dvlden I've released `v0.14.2-alpha2` for testing:

```sh
pnpm i wxt@next
```

**Comment by dvlden** [comment]
Amazing! I will definitely test it tomorrow. It's late now and I'm headed to bed. Giving you my feedback tomorrow, thanks man! 🤙

**Comment by aklinker1** [comment]
Alright, I've integrated the alpha release into one of my extensions to try it out. Everything works great, but there's some DX improvements I'd like to make:

1. I shouldn't have to cast a number to a string when passing it as a substitution:
   ```diff
   -i18n.tp("diffs_additionsText", count, [String(count)])
   +i18n.tp("diffs_additionsText", count, [count])
   ```
2. I don't like having plurals as a separate function. `tp` &rarr; `t`.
   ```diff
   -i18n.tp("diffs_additionsText", count, [count])
   +i18n.t("diffs_additionsText", count, [count])
   ```
3. The nested messages are hard to edit. I can't double click to select one "level" of the message name, like "description1" to quickly change it to "description2", because `_` are not considered word breaks. Same with using `ctrl+backspace` or `opt+delete` to backspace an entire word.
   ```vue
   <p class="font-medium text-base-content">
     {{ t("options_customLists_description1") }}
   </p>
   ```
   I don't think there's anything we can do about this one... Ideally, we would use periods to denote nested levels, like `options.customLists.description1`, but `.` is not a valid character for a message name in web extensions... So we would have to write them to the disk with `_`, and replace all `.` with underscores when grabbing the keys at runtime... I don't want to cause a performance issue when running `.replaceAll(".", "_")` (or a custom `O(n)` loop), but it might be worth it for the convenience...
4. Log a warning and fail the build when a locales directory is present, but no `default_locale` is specified in the manifest, and vise-versa. If we kept the build going, Chrome would fail to load both these extensions, so it's better we catch this sooner.
5. Unlike the `browser.runtime.getMessage` API, I can't hover over the function to see the description and text. It's an easy, built-in way of quickly viewing what the translation text is.
   Screenshot 2024-01-14 at 5 24 12 PM
   Screenshot 2024-01-14 at 5 23 35 PM
6. Substitutions array is typed as `any`, it should be `string[]`

**Comment by dvlden** [comment]
Sorry for slower response. I just grabbed a chance to test this.

Notes (alpha):

- Running without `locales` dir
```
Failed to update .wxt directory: The "path" argument must be of type string or an instance of Buffer or URL. Received undefined
```


- Looks like we are not utilising arrays in locales. Assuming that browser.i18n does not support it natively _(did not research)_
```yaml
# Not working
state:
  - Normal
  - Upscale
  - Stretch
```

- Does not seem to support multiple nesting levels. Not sure why, cause in your example it looks like it does?
```yaml
activity:
  # ...
  state:
    active: On
    inactive: Off
```

```ts
// Shows autocompletion, but renders as empty string
i18n.t('activity_state_on')
```

- Does not trigger hot-reload if `locales/xx.yaml` is changed _(new properties or editing existing ones)_

I think I found it most annoying, not being able to click on the property key to edit the locale message. You said it doesn't let you do it for nested properties, but I can't do it with a non-nested property either. Not sure if I need an extension for that, but native `browser.i18n.getMessage('xxx')` is also not editable through VSCode with CMD + Click on the argument.

**Comment by aklinker1** [comment]
This PR is way out of date, closing in favor of #758. Will be copying over a lot of code from this PR instead of dealing with merge conflicts.

**Pull request closed without merging**